unixfs/hamt/fetcher.go

```diff
@@ -0,0 +1,297 @@
+package hamt
+
+import (
+	"context"
+	"time"
+	//"fmt"
+	//"os"
+
+	logging "gx/ipfs/QmRb5jh8z2E8hMGN2tkvs1yHynUanqnZ3UeKwgN1i9P1F8/go-log"
+	cid "gx/ipfs/QmcZfnkapfECQGcLZaf9B79NRg7cRa9EnZh4LSbkCzwNvY/go-cid"
+	ipld "gx/ipfs/Qme5bWv7wtjUNGsK2BNGVUFPKiuxWrsqrtvYwCLRw8YFES/go-ipld-format"
+)
+
+var log = logging.Logger("hamt")
+
+// fetcher implements a background fether to retrieve missing child
+// shards in large batches.  It attempts to retrieves the missing
+// shards in an order that allow streaming of the complete hamt
+// directory, assuming a depth first traversal.
+type fetcher struct {
+	// note: the fields in this structure should only be accesses by
+	// the 'mainLoop' go routine, all communication should be done via
+	// channels
+
+	ctx   context.Context
+	dserv ipld.DAGService
+
+	reqRes chan *Shard
+	result chan result
+
+	idle bool
+
+	done chan batchJob
+
+	todoFirst *job            // do this job first since we are waiting for its results
+	todo      jobStack        // stack of jobs that still need to be done
+	jobs      map[*Shard]*job // map of all jobs in which the results have not been collected yet
+
+	// stats relevent for streaming the complete hamt directory
+	doneCnt    int // job's done but results not yet retrieved
+	hits       int // job's result already ready, no delay
+	nearMisses int // job currently being worked on, small delay
+	misses     int // job on todo stack but will be done in the next batch, larger delay
+
+	// other useful stats
+	cidCnt int
+
+	start time.Time
+}
+
+// batchSize must be at least as large as the largest number of cids
+// requested in a single job.  For best perforamce it should likely be
+// sligtly larger as jobs are poped from the todo stack in order and a
+// job close to the batchSize could force a very small batch to run.
+// Recommend minimum size: 256 + 64 = 320
+const batchSize = 320
+
+//
+// fetcher public interface
+//
+
+// startFetcher starts a new fetcher in the background
+func startFetcher(ctx context.Context, dserv ipld.DAGService) *fetcher {
+	log.Infof("fetcher: starting...")
+	f := &fetcher{
+		ctx:    ctx,
+		dserv:  dserv,
+		reqRes: make(chan *Shard),
+		result: make(chan result),
+		idle:   true,
+		done:   make(chan batchJob),
+		jobs:   make(map[*Shard]*job),
+	}
+	go f.mainLoop()
+	return f
+}
+
+// result contains the result of a job, see getResult
+type result struct {
+	vals map[string]*Shard
+	errs []error
+}
+
+// get gets the missing child shards for the hamt object.
+// The missing children for the passed in shard is returned.  The
+// children are then also retrieved in the background.  The result is
+// the result of the batch request and not just the single job.  In
+// particular, if the 'errs' field is empty the 'vals' of the result
+// is guaranteed to contain the all the missing child shards, but the
+// map may also contain child shards of other jobs in the batch
```

Comment:
Not sure I understand, so I may get more than I asked for? Or rather, many different `get` operation results may be returned in a single `result` of a particular `get`?